site_scons/site_tools/go_builder.py

```diff
@@ -48,6 +48,17 @@ def _scan_go_file(node, env, _path):
     return includes
 
 
+def get_min_go_version():
+    """Get go minimum version from go.mod"""
+    go_mod_path = os.path.join(Dir('#').abspath, "src", "control", "go.mod")
+    with open(go_mod_path, 'r') as f:
+        for line in f:
+            if line.startswith('go '):  # e.g. "go 1.21"
+                parts = line.split()
+                return parts[1] + ".0"  # e.g. "1.21.0"
```

Comment:
Hmm. I don't think that it's safe to unconditionally append the `.0` suffix here. Per [the spec](https://go.dev/ref/mod#go-mod-file-go), the `go` directive may contain any valid [Go version](https://go.dev/doc/toolchain#version).

After looking at little more closely at the version comparison, I remembered that it doesn't rely on the versions being balanced, i.e. it works fine if one is `x.y.z` and one is `x.y`, so you could just return `parts[1]` here and be done with it, except for the fact that the version might contain an `rc` string. Sigh. You've got the `get_go_version()` helper below that almost does what you need, except that it won't quite match on either the line or just the version. Were it me, I'd probably just do something like `return get_go_version("go" + parts[1])` and call it a day.

The other thing is that if this function somehow fails to get a version from that file, returning `None` will make the version comparison fail with a mysterious error like `AttributeError: 'NoneType' object has no attribute 'split'`. Might be best to just raise a specific exception instead.

Reply:
Good idea. Thanks. I don't know how likely it is for us to use more complicated go version strings in the go.mod file, but simple enough to harden it up.